CPGroupRebalanceTest.test

_master_ (commit 7815ff8a52642ac38b2e75aeaa6c77827f38085c)

Failed on CorrettoJDK11 : http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-CorrettoJDK11/32/testReport/com.hazelcast.cp.internal/CPGroupRebalanceTest/test/

**Stacktrace**
```
java.lang.AssertionError: 
Expected '====== LEADERSHIPS ======
CPMember{uuid=daec1527-dc93-404f-bdde-134e091e2afb, address=[127.0.0.1]:5707} => 11
CPMember{uuid=77de01be-b0a2-49f3-9800-73fe32b399df, address=[127.0.0.1]:5704} => 11
CPMember{uuid=6b1b96c7-6cdb-4459-ba7a-cb92a310dd09, address=[127.0.0.1]:5703} => 11
CPMember{uuid=e366b393-11b8-4f38-9373-11842f975e81, address=[127.0.0.1]:5701} => 10
CPMember{uuid=bfe35401-43b1-4a48-836a-0caf9d6d1bea, address=[127.0.0.1]:5702} => 11
CPMember{uuid=c4428809-02b1-4c0f-93ac-f678f68785d7, address=[127.0.0.1]:5705} => 15
CPMember{uuid=70cf178d-9568-470b-a21f-13dfdad3ef2d, address=[127.0.0.1]:5706} => 11
' to be between 10 and 14, but was 15
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertBetween(HazelcastTestSupport.java:1459)
	at com.hazelcast.cp.internal.CPGroupRebalanceTest.test(CPGroupRebalanceTest.java:73)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Another failure
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/4828/testReport/junit/com.hazelcast.cp.internal/CPGroupRebalanceTest/test/

Error Message
```
Expected '====== LEADERSHIPS ======
CPMember{uuid=c1e0f34f-73c3-45f7-bb52-45537c9bdfdf, address=[127.0.0.1]:5704} => 11
CPMember{uuid=62ece9ed-61e8-4033-87cf-33e777e418a7, address=[127.0.0.1]:5707} => 11
CPMember{uuid=404a2c7d-1c4e-4d8d-8400-85745874c440, address=[127.0.0.1]:5706} => 15
CPMember{uuid=96300c8f-5a9f-4109-b8ea-b06d1ae10434, address=[127.0.0.1]:5703} => 11
CPMember{uuid=5a3cea45-05bc-45b9-8c6f-457511daf553, address=[127.0.0.1]:5702} => 10
CPMember{uuid=28fd18fb-112b-4d37-b8a1-92bc0cbe170d, address=[127.0.0.1]:5701} => 11
CPMember{uuid=62daab0b-cbb4-4b54-b5f8-7d4daa759d44, address=[127.0.0.1]:5705} => 11
' to be between 10 and 14, but was 15
```
Stacktrace
```
java.lang.AssertionError: 
Expected '====== LEADERSHIPS ======
CPMember{uuid=c1e0f34f-73c3-45f7-bb52-45537c9bdfdf, address=[127.0.0.1]:5704} => 11
CPMember{uuid=62ece9ed-61e8-4033-87cf-33e777e418a7, address=[127.0.0.1]:5707} => 11
CPMember{uuid=404a2c7d-1c4e-4d8d-8400-85745874c440, address=[127.0.0.1]:5706} => 15
CPMember{uuid=96300c8f-5a9f-4109-b8ea-b06d1ae10434, address=[127.0.0.1]:5703} => 11
CPMember{uuid=5a3cea45-05bc-45b9-8c6f-457511daf553, address=[127.0.0.1]:5702} => 10
CPMember{uuid=28fd18fb-112b-4d37-b8a1-92bc0cbe170d, address=[127.0.0.1]:5701} => 11
CPMember{uuid=62daab0b-cbb4-4b54-b5f8-7d4daa759d44, address=[127.0.0.1]:5705} => 11
' to be between 10 and 14, but was 15
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertBetween(HazelcastTestSupport.java:1488)
	at com.hazelcast.cp.internal.CPGroupRebalanceTest.test(CPGroupRebalanceTest.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/CPGroupRebalanceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/CPGroupRebalanceTest.java
@@ -21,6 +21,8 @@ import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.cp.CPGroupId;
 import com.hazelcast.cp.CPMember;
 import com.hazelcast.cp.internal.operation.GetLeadedGroupsOp;
+import com.hazelcast.cp.internal.raft.impl.RaftEndpoint;
+import com.hazelcast.cp.internal.raftop.metadata.CreateRaftGroupOp;
 import com.hazelcast.spi.impl.InternalCompletableFuture;
 import com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl;
 import com.hazelcast.test.HazelcastSerialClassRunner;
@@ -31,11 +33,14 @@ import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 
+import static com.hazelcast.cp.internal.MetadataRaftGroupManager.INITIAL_METADATA_GROUP_ID;
 import static com.hazelcast.cp.internal.RaftGroupMembershipManager.LEADERSHIP_BALANCE_TASK_PERIOD;
 
 @RunWith(HazelcastSerialClassRunner.class)
@@ -52,14 +57,15 @@ public class CPGroupRebalanceTest extends HazelcastRaftTestSupport {
     @Test
     public void test() throws Exception {
         int cpMemberCount = 7;
-        HazelcastInstance[] instances = newInstances(cpMemberCount, 3, 0);
+        int groupSize = 3;
+        HazelcastInstance[] instances = newInstances(cpMemberCount, groupSize, 0);
         waitUntilCPDiscoveryCompleted(instances);
 
-        final int leadershipsPerMember = 11;
-        final int extraGroups = 3;
-        final int groupCount = cpMemberCount * leadershipsPerMember + extraGroups;
+        int leadershipsPerMember = 11;
+        int extraGroups = 3;
+        int groupCount = cpMemberCount * leadershipsPerMember + extraGroups;
 
-        createRaftGroups(instances, groupCount);
+        createRaftGroups(instances, groupSize, groupCount);
 
         HazelcastInstance metadataLeader = getLeaderInstance(instances, getMetadataGroupId(instances[0]));
         Collection<CPMember> cpMembers = metadataLeader.getCPSubsystem().getCPSubsystemManagementService().getCPMembers()
@@ -79,20 +85,33 @@ public class CPGroupRebalanceTest extends HazelcastRaftTestSupport {
         getRaftService(metadataLeader).getMetadataGroupManager().rebalanceGroupLeaderships();
     }
 
-    private void createRaftGroups(HazelcastInstance[] instances, int groupCount) {
+    private void createRaftGroups(HazelcastInstance[] instances, int groupSize, int groupCount) {
         RaftInvocationManager invocationManager = getRaftInvocationManager(instances[0]);
 
+        RaftEndpoint[] endpoints = Arrays.stream(instances)
+                                         .map(instance -> instance.getCPSubsystem().getLocalCPMember())
+                                         .map(cpMember -> ((CPMemberInfo) cpMember).toRaftEndpoint())
+                                         .toArray(RaftEndpoint[]::new);
         Collection<CPGroupId> groupIds = new ArrayList<>(groupCount);
-        Collection<InternalCompletableFuture<RaftGroupId>> futures = new ArrayList<>(groupCount);
+        Collection<InternalCompletableFuture<CPGroupSummary>> futures = new ArrayList<>(groupCount);
 
         for (int i = 0; i < groupCount; i++) {
-            InternalCompletableFuture<RaftGroupId> f = invocationManager.createRaftGroup("group-" + i);
+            List<RaftEndpoint> groupMembers = new ArrayList<>(groupSize);
+            for (int j = 0; j < groupSize; j++) {
+                groupMembers.add(endpoints[(i + j) % endpoints.length]);
+            }
+
+            RaftOp op = new CreateRaftGroupOp("group-" + i, groupMembers);
+            InternalCompletableFuture<CPGroupSummary> f = invocationManager.invoke(INITIAL_METADATA_GROUP_ID, op);
             futures.add(f);
         }
-        for (InternalCompletableFuture<RaftGroupId> future : futures) {
-            RaftGroupId groupId = future.join();
-            groupIds.add(groupId);
+
+        for (InternalCompletableFuture<CPGroupSummary> future : futures) {
+            CPGroupSummary group = future.join();
+            invocationManager.triggerRaftNodeCreation(group);
+            groupIds.add(group.id());
         }
+
         for (CPGroupId groupId : groupIds) {
             // await leader election
             getLeaderInstance(instances, groupId);

